compiler/src/pretyping.ml

```diff
@@ -1812,31 +1819,45 @@ let rec tt_instr arch_info (env : 'asm Env.env) ((annot,pi) : S.pinstr) : 'asm E
       if List.is_empty ls then raise exn;
       let pe1 = pexpr_of_plvalue exn (List.last ls) in
       let pe  = L.mk_loc loc (S.PEOp2(op,(pe1,pe))) in
-      tt_assign env (pimp, ls) `Raw pe None
+      tt_assign env_lhs env_rhs (pimp, ls) `Raw pe None
 
   | ls, eqop, e, Some cp ->
       let loc = L.loc pi in
       let exn = Unsupported "if not allowed here" in
-      let env, i = tt_assign env ls eqop e None in
+      let i = tt_assign env_lhs env_rhs ls eqop e None in
       let x, ty, e, is =
         match i with
         | { i_desc = P.Cassgn (x, _, ty, e) ; _ } :: is -> x, ty, e, is
         | _ -> rs_tyerror ~loc exn in
       let e' = oget ~exn:(tyerror ~loc exn) (P.expr_of_lval x) in
-      let c = tt_expr_bool arch_info.pd env cp in
-      env, mk_i (P.Cassgn (x, AT_none, ty, Pif (ty, c, e, e'))) :: is
+      let c = tt_expr_bool arch_info.pd env_rhs cp in
+      mk_i (P.Cassgn (x, AT_none, ty, Pif (ty, c, e, e'))) :: is
   in
+  let tt_annot_decl env (vd: S.vardecl L.located) (aty: A.annotations * S.pstotype) =
+    (* remember the environment prior to the declaration:
+      it will be used to type-check the initializing expression, if any *)
+    let old_env = env in
```

Comment:
maybe old_env can be renamed in env_rhs and env in env_lhs.
But it is no important